clang/include/clang/ExtractAPI/Serialization/SerializerBase.h

```diff
@@ -28,11 +28,78 @@ struct APISerializerOption {
 };
 
 /// The base interface of serializers for API information.
-class APISerializer {
+template <typename Derived> class APISerializer {
 public:
   /// Serialize the API information to \p os.
   virtual void serialize(raw_ostream &os) = 0;
 
+  virtual void traverseGlobalFunctionRecords() {
+    for (const auto &globalFunction : API.getGlobalFunctions())
+      static_cast<Derived *>(this)->visitGlobalFunctionRecord(
+          std::move(*globalFunction.second));
+  }
+
+  void traverseGlobalVariableRecords() {
+    for (const auto &global : API.getGlobalVariables())
+      static_cast<Derived *>(this)->visitGlobalVariableRecord(*global.second);
```

Comment:
It might be nice to refactor `static_cast<Derived *>(this)` into a protected method `Derived *getDerived() { return static_cast<Derived *>(this); }` to aid readability.